### ISSUE TITLE
Segmentation fault in BrainSpeechDecoder::Rescore()

### ISSUE THREAD

**Issue by donatojb** [fwillett/speechBCI#13]
I am running [this notebook](https://github.com/fwillett/speechBCI/blob/49a950772ab732e656ef3f03eb91b9cfbd2333ae/AnalysisExamples/5gram%2Bllm_rescoring.ipynb) and I got a segmentation fault in the N-gram decoding loop. The line that is triggering the segfault is [this one](https://github.com/fwillett/speechBCI/blob/49a950772ab732e656ef3f03eb91b9cfbd2333ae/NeuralDecoder/neuralDecoder/utils/lmDecoderUtils.py#L190C26-L190C26). I downloaded the N-gram graphs from [here](https://datadryad.org/stash/dataset/doi:10.5061/dryad.x69p8czpq) and compiled following [these instructions](https://github.com/fwillett/speechBCI/blob/49a950772ab732e656ef3f03eb91b9cfbd2333ae/LanguageModelDecoder/README.md). Running the code with `rescore=False` when calling `lmDecoderUtils.lm_decode` works well.

This example produces the segmentation fault: 

```
import neuralDecoder.utils.lmDecoderUtils as lmDecoderUtils

lmDir = './languageModel'
ngramDecoder = lmDecoderUtils.build_lm_decoder(
    lmDir,
    acoustic_scale=0.5,
    nbest=100,
    beam=18
)

ngramDecoder.Rescore()
```

The otuput in the terminal is the following:
```
WARNING: Logging before InitGoogleLogging() is written to STDERR
I0117 12:34:17.155279 762859 brain_speech_decoder.h:52] Reading fst ./languageModel/TLG.fst
I0117 12:36:37.021819 762859 brain_speech_decoder.h:81] Reading symbol table ./languageModel/words.txt
Fatal Python error: Segmentation fault

Current thread 0x00007ff887cfb740 (most recent call first):
  File "/home/gridsan/dbeneto/TFG/speechBCI/debug.py", line 11 in <module>
Segmentation fault
```

**Comment by cffan** [comment]
Thanks for all the info. We'll look into this soon. One question: did you use the pretrained RNN or your custom model?

**Comment by donatojb** [comment]
I used the pretrained RNN

**Comment by cffan** [comment]
I couldn't reproduce the segfault in this [notebook](https://github.com/fwillett/speechBCI/blob/49a950772ab732e656ef3f03eb91b9cfbd2333ae/AnalysisExamples/5gram%2Bllm_rescoring.ipynb). One thing I noticed is that if `lmDecoderUtils.lm_decode` is interrupted while running, it may crash the next time it runs because corrupted internal state.

The segfault in the example code you provided is because you cannot call `Rescore` before running the actually decoding. The correct code should be:
```
lm_decoder.DecodeNumpy(ngramDecoder, logits, logPriors, blankPenalty)
ngramDecoder.FinishDecoding()
ngramDecoder.Rescore()
```

Can you also list what's inside your `languageModel` folder?

**Comment by LeonHermann322** [comment]
I have the same problem at the moment. If I'm deactivating rescore it executes as expected, but with rescore active it fails during the rescoring. 
The language model was downloaded from the most recent language model [here](https://datadryad.org/stash/dataset/doi:10.5061/dryad.x69p8czpq). Its contents are
`G.fst  L.fst  LG.fst  T.fst  TLG.fst  lexicon_numbers.txt  tokens.txt  units.txt  words.txt`.
I'm executing this [script](https://github.com/cffan/neural_seq_decoder/blob/master/scripts/eval_competition.py) for evaluation. 

> lm_decoder.DecodeNumpy(ngramDecoder, logits, logPriors, blankPenalty)
ngramDecoder.FinishDecoding()
ngramDecoder.Rescore()

The execution order as mentioned in your previous comment should be correct as this script is executing the function [neuralDecoder.utils.lmDecoderUtils.lm_decode](https://github.com/fwillett/speechBCI/blob/49a950772ab732e656ef3f03eb91b9cfbd2333ae/NeuralDecoder/neuralDecoder/utils/lmDecoderUtils.py) for this.

**Comment by cffan** [comment]
@LeonHermann322, did you use the pre-trained RNN or your custom model?

**Comment by LeonHermann322** [comment]
I'm using a custom model, but the model output should be structured in the same way. Since @donatojb was using the pretrained RNN I assumed that there could be a bug independent from the model. Right now, I'm working on making sure that the error is not caused by our model output.

**Comment by donatojb** [comment]
I was also using the 3gram LM, so that was the problem, thank you. Is there a download link for the 5gram model?

**Comment by LeonHermann322** [comment]
Ah, that has to be it then. Thank you for your replies!

**Comment by cffan** [comment]
@donatojb, you can download the 5gram LM from the same website where you downloadd the 3gram LM. 

![Uploading Screenshot 2024-03-19 at 12.21.15 PM.png…]()
